assets/tests/components/routeLadder.test.tsx

```diff
@@ -0,0 +1,114 @@
+import React from "react"
+import renderer, { act } from "react-test-renderer"
+import RouteLadder from "../../src/components/routeLadder"
+import { Route } from "../../src/skate"
+import { TimepointsForRouteId } from "../../src/state"
+
+declare global {
+  interface Window {
+    /* eslint-disable typescript/no-explicit-any */
+    fetch: (uri: string) => Promise<any>
+  }
+}
+
+const dispatch = () => undefined
+
+test("renders a route ladder", () => {
+  const route: Route = { id: "28" }
+  const timepointsForRouteId: TimepointsForRouteId = () => [
+    { id: "MATPN" },
+    { id: "WELLH" },
+    { id: "MORTN" },
+  ]
+
+  const tree = renderer
+    .create(
+      <RouteLadder
+        route={route}
+        dispatch={dispatch}
+        timepointsForRouteId={timepointsForRouteId}
+      />
+    )
+    .toJSON()
+
+  expect(tree).toMatchSnapshot()
+})
+
+test("displays loading if we are fetching the timepoints", () => {
+  const route: Route = { id: "28" }
+  const timepointsForRouteId: TimepointsForRouteId = () => null
+
+  const tree = renderer
+    .create(
+      <RouteLadder
+        route={route}
+        dispatch={dispatch}
+        timepointsForRouteId={timepointsForRouteId}
+      />
+    )
+    .toJSON()
+
+  expect(tree).toMatchSnapshot()
+})
+
+test("fetches timepoints for this route if we don't yet have them", () => {
+  const mockDispatch = jest.fn()
+  const route: Route = { id: "28" }
+  const timepointsForRouteId: TimepointsForRouteId = () => undefined
+
+  window.fetch = () =>
+    Promise.resolve({
+      json: () => ({
+        data: [{ id: "28" }, { id: "39" }, { id: "71" }],
+      }),
+      ok: true,
+      status: 200,
+    })
+
+  act(() => {
+    renderer
+      .create(
+        <RouteLadder
+          route={route}
+          dispatch={mockDispatch}
+          timepointsForRouteId={timepointsForRouteId}
+        />
+      )
+      .toJSON()
+  })
+
+  expect(mockDispatch.mock.calls.length).toBe(1)
```

Comment:
Shouldn't we have two calls? One to say we're loading, and a second one once the api call returns?

Reply:
The second one happens asynchronously so the test finishes before that happens. But I tested the reducer to test handling that data.

Reply:
If the promise doesn't resolve, before the test finishes, does its return value matter?

Is there a way to make the test wait for the promise? This page has some options: https://jestjs.io/docs/en/asynchronous

Reply:
Trouble is we aren't calling the promise, we are triggering it by interacting with the DOM. But I think demostrating that the fetch process starts seems like a sufficient test here to me.